### PR TITLE
[geoclue-provider-hybris] Add Cold Start. Contributes to JB#41709

### PR DESCRIPTION
**Pull request by dmamaevomp**
Added org.freedesktop.Geoclue.Data interface to support GNSS cold start.

### CODE REVIEW

File: org.freedesktop.Geoclue.Data.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN" "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="org.freedesktop.Geoclue.Data">
```
Comment (pvuorela) [inline org.freedesktop.Geoclue.Data.xml]: This would be your own api? If so, why call it geoclue? Could be some hybrisprovider specific?

Comment (dmamaevomp) [reply org.freedesktop.Geoclue.Data.xml]: GPS is part of the geoclue. cold start is part of the GPS. Accordingly, this should relate directly to geoclue.

Comment (pvuorela) [reply org.freedesktop.Geoclue.Data.xml]: You seem to be anyway introducing a new api which is hybris specific. It's not part of geoclue, just an extra interface on a geoclue provider.

Furthermore you've multiple times suggested to use the common options api way which doesn't need any new interface, yet no rationale given why it wouldn't be good.

Comment (dmamaevomp) [reply org.freedesktop.Geoclue.Data.xml]: Please check the updated code, I added a new interface com.omp.gps.service

File: hybrisprovider.cpp [hunk 2]
```diff
@@ -199,6 +199,10 @@ HybrisProvider::HybrisProvider(QObject *parent)
         m_idleTimer.start(QuitIdleTime, this);
     }
 
+    m_dbusComOmpGpsService = new DBusComOmpGpsService();
```
Comment (pvuorela) [inline hybrisprovider.cpp]: Please don't call this omp, these things shouldn't be connected to companies or such. Also it's not necessarily "the" gps service even without the omp part.

Earlier suggested suggested using hybris provider specific naming, that still holds.

Then again I'm not sure does this even need a separate _service_, or would it suffice just providing the functionality under some separate interface. There's already org.freedesktop.Geoclue.Providers.Hybris registered for this.

Comment (dmamaevomp) [reply hybrisprovider.cpp]: I used the names of the company, so the code used the names jolla and jollamobile earlier. Now I have removed the name of the company. 
I initially suggested that you add this to Geoclue, but Chris Adams suggested excluding org.sailfishos.geoclue.gnss from geoclue and adding a new interface. 
Perhaps I misunderstood your comments, but then please describe in detail what and how to do. 
Thanks in advance!

**Comment by pvuorela** [comment]
No, it shouldn't be com.gps either, that's overly generic.

You didn't comment anything if this even needs a separate service.

And you haven't still commented anything about the common api.

File: hybrisprovider.cpp (hunk 2 above)
Comment (dmamaevomp) [reply hybrisprovider.cpp]: For example, 
1) add org.freedesktop.Geoclue.Providers.Hybris.GPS.xml
2) add the GnssDeleteAidingData method to org.freedesktop.Geoclue.Providers.Hybris.GPS.xml 
3) and etc

Comment (dmamaevomp) [reply hybrisprovider.cpp]: Еhis method is only used when testing the gps receiver. It is enough to call it only once and periodic use is not planned

**Comment by chriadam** [comment]
Thanks for your work on this, I think we're definitely getting closer to merge-able state.

My personal opinion is that:

1) we should favour using the SetOptions() way if that will suffice, as then we don't need to add or extend any interface, but can simply use the existing one
2) if the SetOptions() way won't suffice, and a custom interface is needed, then I would prefer the interface to be org.freedesktop.Geoclue.Providers.Hybris as per Pekka's earlier suggestion

**Comment by dmamaevomp** [comment]
> Thanks for your work on this...
How do I use the SetOptions method? I didn't quite understand. Please explain

**Comment by chriadam** [comment]
> > Thanks for your work on this...
> 
> How do I use the SetOptions method? I didn't quite understand. Please explain

My understanding is that you just need to pass a QVariantMap (i.e. DBus a{sv}) parameter.
e.g. clients could call SetOptions({ "GnssDeleteAidingData": true })

The code in ::SetOptions() would have some extra check like:

```
if (options.contains(QStringLiteral("GnssDeleteAidingData"))
        && options.value(QStringLiteral("GnssDeleteAidingData")).toBool()
        && m_backend) {
    //GPS_DELETE_ALL = 0xFFFF (almanac, ephemeris, position, time and other cache data)
    m_backend->gnssDeleteAidingData(0xFFFF);
}
```

You can see that SetOptions is already used to set the UpdateInterval (i.e. polling interval) for the plugin, currently.

--

After thinking about this a bit more, it does seem like these Options are more like settings rather than intended for a way to trigger some behaviour (such as this GnssDeleteAidingData).  So perhaps a separate interface with bespoke invokable method is the best option after all: @pvuorela what's your thoughts currently?

**Comment by dmamaevomp** [comment]
Thanks for the clarification. Now everything is clear to me!
I agree that this is not an option, but I do not mind doing it through SetOptions
Then I wait for the final decision from @pvuorela

**Comment by pvuorela** [comment]
> After thinking about this a bit more, it does seem like these Options are more like settings rather than intended for a way to trigger some behaviour (such as this GnssDeleteAidingData). So perhaps a separate interface with bespoke invokable method is the best option after all: @pvuorela what's your thoughts currently?

True, it's not a search option as much as some others, for example that "UpdateInterval". Then again if wanted it could be named a bit more like one, say "NoCachedAidingData".

But if it's now really just needed for testing help, I think the options way would be the most straightforward. One thing I could consider if the options name should include hybris to make it really provider specific. Like "HybrisDeleteAidingData".

**Comment by dmamaevomp** [comment]
Please check the updated code

**Comment by pvuorela** [comment]
This I can approve, thanks!

LGTM.

**Comment by dmamaevomp** [comment]
Thank you all!

**Pull request merged**